### PR TITLE
Track the latest changes from upstream rustls

### PR DESCRIPTION
**Pull request by stevefan1999-personal**

### CODE REVIEW

File: Cargo.toml [hunk 1]
```diff
@@ -14,21 +14,23 @@ exclude = ["/.github", "/examples", "/scripts"]
 
 [dependencies]
 tokio = "1.0"
-rustls = { version = "=0.22.0-alpha.2", default-features = false }
+rustls = { version = "0.22.0-alpha.2", default-features = false }
```
Comment (djc) [inline Cargo.toml:17]: We want to stick to using `=` for semver pre-releases, because Cargo otherwise treats pre-releases as compatible even though they often aren't. If you want to pull in the latest rustls changes, please submit a version bump PR in the rustls repo to request another alpha.

File: Cargo.toml [hunk 2]
```diff
@@ -14,21 +14,23 @@ exclude = ["/.github", "/examples", "/scripts"]
...
 default = ["logging", "tls12"]
-dangerous_configuration = ["rustls/dangerous_configuration"]
 early-data = []
 logging = ["rustls/logging"]
 secret_extraction = ["rustls/secret_extraction"]
 tls12 = ["rustls/tls12"]
 
 [dev-dependencies]
-argh = "0.1"
+argh = "0.1.1"
```
Comment (djc) [inline Cargo.toml:27]: This change is unrelated and seems unnecessary.

File: Cargo.toml [hunk 3]
```diff
@@ -14,21 +14,23 @@ exclude = ["/.github", "/examples", "/scripts"]
...
 futures-util = "0.3.1"
-lazy_static = "1"
+lazy_static = "1.1"
 webpki-roots = "=0.26.0-alpha.1"
 rustls-pemfile = "=2.0.0-alpha.1"
-webpki = { package = "rustls-webpki", version = "=0.102.0-alpha.2", features = ["alloc", "std"] }
+webpki = { package = "rustls-webpki", version = "=0.102.0-alpha.3", features = ["alloc", "std"] }
+
+[patch.crates-io]
+rustls = { git = 'https://github.com/rustls/rustls' }
```
Comment (djc) [inline Cargo.toml:36]: This is unused.

File: Cargo.toml [hunk 4]
```diff
@@ -14,21 +14,23 @@ exclude = ["/.github", "/examples", "/scripts"]
...
-argh = "0.1"
+argh = "0.1.1"
 tokio = { version = "1.0", features = ["full"] }
 futures-util = "0.3.1"
-lazy_static = "1"
+lazy_static = "1.1"
 webpki-roots = "=0.26.0-alpha.1"
 rustls-pemfile = "=2.0.0-alpha.1"
-webpki = { package = "rustls-webpki", version = "=0.102.0-alpha.2", features = ["alloc", "std"] }
+webpki = { package = "rustls-webpki", version = "=0.102.0-alpha.3", features = ["alloc", "std"] }
```
Comment (djc) [inline Cargo.toml:33]: Please submit one specific commit (or if you have don't have experience wrangling commits, one PR), per logical change.

File: Cargo.toml (hunk 3 above)
Comment (stevefan1999-personal) [reply Cargo.toml:36]: Well, by tracking the latest changes from upstream rustls I mean I'm really using the version from main branch. So without the patch, that means we are going to use the crates.io version of rustls ([rustls - crates.io: Rust Package Registry](https://crates.io/crates/rustls/0.22.0-alpha.2)) which is not exactly the intent

File: Cargo.toml (hunk 4 above)
Comment (stevefan1999-personal) [reply Cargo.toml:33]: Reason ditto for the rustls patch change because of this (https://github.com/rustls/rustls/blob/86aa146d59f83e9fbf890b05f823799c0e8d3023/rustls/Cargo.toml#L22C5-L22C5). The intent is to track the latest main branch, and not really going after the crates.io version.

File: Cargo.toml (hunk 3 above)
Comment (djc) [reply Cargo.toml:36]: Ah, sorry, you're right. Anyway, I don't want to track a branch here without pinning a commit, and I'd prefer just publishing another rustls alpha.

Comment (stevefan1999-personal) [reply Cargo.toml:36]: sure, I will remove the patch and pin the right version when the next rustls alpha is cooked on crates.io.

**Review by djc — APPROVED** [review]

**Comment by djc** [comment]
Let's get this PR merged (after it passes CI first), then do a separate PR for bumping the version number.

**Comment by stevefan1999-personal** [comment]
Hmm...should we have enabled ring feature by default?

**Comment by djc** [comment]
> Hmm...should we enable ring feature by default?

We should match rustls in this regard.

**Review by djc — APPROVED** [review]
